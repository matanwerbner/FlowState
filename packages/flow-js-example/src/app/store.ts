import { create } from 'zustand';
import {
  InitialData,
  InitialState,
  createStateMachine,
} from './lib/createStateMachine';
import { EventNames, GameContext, GameStates } from './lib/types';

type InitializeParams<GameContext> = {
  state?: GameStates;
  data?: Partial<GameContext>;
};

type Storetype = {
  currentState: GameStates;
  triggerEvent: (eventName: EventNames, data?: Partial<GameContext>) => void;
  context: GameContext;
  initializeStore: ({ state, data }: InitializeParams<GameContext>) => void;
};

export const useGameStore = create<Storetype>((set) => {
  const stateMachine = createStateMachine();

  return {
    context: stateMachine.context,
    currentState: stateMachine.currentState,
    triggerEvent: (eventName: EventNames, data?: Partial<GameContext>) => {
      return stateMachine.triggerEvent(eventName, data).then((newState) => {
        set({ currentState: newState, context: stateMachine.context });
      });
    },
    initializeStore: ({ state = InitialState, data = InitialData }) => {
      stateMachine.initialize(state, { ...InitialData, ...data });
      set({
        currentState: stateMachine.currentState,
        context: stateMachine.context,
      });
    },
  };
});
