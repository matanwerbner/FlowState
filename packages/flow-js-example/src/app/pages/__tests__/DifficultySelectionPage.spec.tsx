import { useGameStore } from '../../store';
import { render, fireEvent } from '@testing-library/react';
import { DifficultySelectionPage } from '../DifficultySelectionPage.tsx';

describe('DifficultySelectionPage tests', () => {
  it('should render', () => {
    const { baseElement } = render(<DifficultySelectionPage />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should trigger event on change', () => {
    const triggerEvent = jest.fn();
    useGameStore.setState({ triggerEvent });
    const { getByTestId } = render(<DifficultySelectionPage />);
    fireEvent.change(getByTestId('difficulty-select'), {
      target: { value: 'easy' },
    });
    expect(triggerEvent).toHaveBeenCalledWith('difficultyPicked', {
      difficulty: 'easy',
    });
  });
});
