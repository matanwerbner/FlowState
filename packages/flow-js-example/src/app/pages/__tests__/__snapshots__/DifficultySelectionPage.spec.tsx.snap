// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DifficultySelectionPage tests should render 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
      >
        Questions difficulty
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      >
        <div
          aria-controls=":r0:"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          role="combobox"
          tabindex="0"
        >
          Choose Difficulty
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          data-testid="difficulty-select"
          tabindex="-1"
          value="-1"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-ihdtdm"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</body>
`;
